### PR TITLE
Convert lock APIs to Universal and fix inconsistencies

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
This is another API conversion to Universal (single implementation of business logic in gRPC, that is exposed as HTTP API too), with the usual goals:

- Have less APIs implemented in HTTP so the migration to net/http will be simpler (#4979)
- Have less code to maintain in the future
- Ensure gRPC and HTTP APIs are consistent
  - In this case, the HTTP APIs were using Resiliency, while gRPC ones were not. They both use Resiliency now.

### CODE REVIEW

File: pkg/grpc/universalapi/api_lock.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	"github.com/dapr/components-contrib/lock"
+	lockLoader "github.com/dapr/dapr/pkg/components/lock"
+	"github.com/dapr/dapr/pkg/messages"
+	runtimev1pb "github.com/dapr/dapr/pkg/proto/runtime/v1"
+	"github.com/dapr/dapr/pkg/resiliency"
+)
+
+func (a *UniversalAPI) TryLockAlpha1(ctx context.Context, req *runtimev1pb.TryLockRequest) (*runtimev1pb.TryLockResponse, error) {
+	// 1. validate and find lock component
+	if req.ExpiryInSeconds <= 0 {
```
Comment (mukundansundar) [inline pkg/grpc/universalapi/api_lock.go:28]: completely unrelated to this PR... ideally having data on the overhead i.e. perf numbers for this API is a must before making it stable.

File: pkg/http/api_test.go [hunk 2]
```diff
@@ -2894,14 +2898,14 @@ func TestV1Alpha1DistributedLock(t *testing.T) {
 		b, _ := json.Marshal(&req)
 
 		resp := fakeServer.DoRequest("POST", apiPath, b, nil)
-		assert.Equal(t, 500, resp.StatusCode)
+		assert.Equal(t, 400, resp.StatusCode)
```
Comment (mukundansundar) [inline pkg/http/api_test.go:2901]: should this be `403 Forbidden` instead?

File: pkg/http/api_test.go [hunk 3]
```diff
@@ -2876,14 +2880,14 @@ func TestV1Alpha1DistributedLock(t *testing.T) {
...
```
Comment (mukundansundar) [inline pkg/http/api_test.go:2883]: should this be `404` instead?

File: pkg/http/api_test.go [hunk 4]
```diff
@@ -2948,17 +2952,15 @@ func TestV1Alpha1DistributedLock(t *testing.T) {
...
 		// assert
-		assert.NotNil(t, resp.JSONBody)
-		rspMap := resp.JSONBody.(map[string]interface{})
-		assert.NotNil(t, rspMap)
-		assert.Equal(t, float64(3), rspMap["status"])
+		assert.Contains(t, string(resp.RawBody), "ERR_MALFORMED_REQUEST")
+		assert.Contains(t, string(resp.RawBody), "ResourceId is empty in lock store store1")
 	})
 
 	t.Run("Unlock with invalid resource id that returns 500", func(t *testing.T) {
```
Comment (mukundansundar) [inline pkg/http/api_test.go:2962]: shouldn't this also return `404`?

File: pkg/messages/predefined.go [hunk 5]
```diff
@@ -111,6 +104,15 @@ var (
...
 
+	// Lock.
+	ErrLockStoresNotConfigured    = APIError{"lock store is not configured", "ERR_LOCK_STORE_NOT_CONFIGURED", http.StatusInternalServerError, grpcCodes.FailedPrecondition}
+	ErrResourceIDEmpty            = APIError{"ResourceId is empty in lock store %s", "ERR_MALFORMED_REQUEST", http.StatusBadRequest, grpcCodes.InvalidArgument}
+	ErrLockOwnerEmpty             = APIError{"LockOwner is empty in lock store %s", "ERR_MALFORMED_REQUEST", http.StatusBadRequest, grpcCodes.InvalidArgument}
+	ErrExpiryInSecondsNotPositive = APIError{"ExpiryInSeconds is not positive in lock store %s", "ERR_MALFORMED_REQUEST", http.StatusBadRequest, grpcCodes.InvalidArgument}
+	ErrLockStoreNotFound          = APIError{"lock store %s not found", "ERR_LOCK_STORE_NOT_FOUND", http.StatusBadRequest, grpcCodes.InvalidArgument}
+	ErrTryLockFailed              = APIError{"failed to try acquiring lock: %s", "ERR_TRY_LOCK", http.StatusInternalServerError, grpcCodes.Internal}
+	ErrUnlockFailed               = APIError{"failed to release lock: %s", "ERR_UNLOCK", http.StatusInternalServerError, grpcCodes.Internal}
+
```
Comment (mukundansundar) [inline pkg/messages/predefined.go:115]: why are all http status codes bad request or ISE?
same for gRPC, we can have corresponding codes?

**Comment by mukundansundar** [comment]
Also a breaking change with the existing PR HTTP status code changes?

File: pkg/http/api_test.go (hunk 2 above)
Comment (ItalyPaleAle) [reply pkg/http/api_test.go:2901]: Agree not 400, but maybe 409 (Conflict)

File: pkg/http/api_test.go [hunk 6]
```diff
@@ -2911,14 +2915,14 @@ func TestV1Alpha1DistributedLock(t *testing.T) {
 		b, _ := json.Marshal(&req)
 
 		resp := fakeServer.DoRequest("POST", apiPath, b, nil)
-		assert.Equal(t, 500, resp.StatusCode)
+		assert.Equal(t, 400, resp.StatusCode)
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/http/api_test.go:2918]: For invalid data, 400 seems appropriate.

File: pkg/http/api_test.go (hunk 3 above)
Comment (mukundansundar) [reply pkg/http/api_test.go:2883]: If it's a missing parameter I agree... Please do update the test name

File: pkg/http/api_test.go (hunk 2 above)
Comment (mukundansundar) [reply pkg/http/api_test.go:2901]: Isn't 409 more related to version conflicts? 
In this case it's someone else owning the lock right?

File: pkg/http/api_test.go (hunk 4 above)
Comment (ItalyPaleAle) [reply pkg/http/api_test.go:2962]: This was in the original API design, and I didn't change it.

I agree with you that this is not the correct status code, but changing that would require reworking the entire APIs down to the components. Right now, they return an error in all cases where a lock can't be released.

I'll open an issue to track that.

Comment (ItalyPaleAle) [reply pkg/http/api_test.go:2962]: https://github.com/dapr/dapr/issues/3549#issuecomment-1487129276

File: pkg/http/api_test.go [hunk 7]
```diff
@@ -2975,11 +2977,11 @@ func TestV1Alpha1DistributedLock(t *testing.T) {
 	})
 
 	t.Run("Unlock with invalid owner", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/http/api_test.go:2979]: Same as above. I did not change this and doing this change is a fairly large breaking thing (and requires changing code down the stack).

**Comment by ItalyPaleAle** [comment]
@mukundansundar Per comments above, I agree that the status codes right now aren't always perfect. For example, TryLock and Unlock do not distinguish between a lock that doesn't exist or an internal error. The original APIs returned an ISE in both cases. 

I have tracked that here: https://github.com/dapr/dapr/issues/3549#issuecomment-1487129276

I think that those should be addressed as part of the efforts to stabilize this API, but likely out of scope here as they even require changes in components-contrib to return the correct errors.

Most status codes here haven't changed, with some rare exceptions for bad requests. I don't think that's a breaking change we should be concerned with much because they're still error status codes, and these APIs aren't stable anyways.

**Comment by mukundansundar** [comment]
> Most status codes here haven't changed, with some rare exceptions for bad requests. I don't think that's a breaking change we should be concerned with much because they're still error status codes, and these APIs aren't stable anyways.

Are there SDK changes needed for any specific request that have status code changes?

**Review by mukundansundar — APPROVED** [review]

**Comment by mukundansundar** [comment]
> I believe most (all?) SDKs use gRPC, so this wouldn't impact them. But I'm not sure.

Seeing that no handcrafted SDK for now have HTTP APIs implemented.

**Comment by yaron2** [comment]
> > I believe most (all?) SDKs use gRPC, so this wouldn't impact them. But I'm not sure.
> 
> Seeing that no handcrafted SDK for now have HTTP APIs implemented.

PHP does

**Comment by mukundansundar** [comment]
> > > I believe most (all?) SDKs use gRPC, so this wouldn't impact them. But I'm not sure.
> > 
> > 
> > Seeing that no handcrafted SDK for now have HTTP APIs implemented.
> 
> PHP does

specifically I am not seeing a lock api implementation in PHP SDK.

**Pull request merged**